File: lib/g3/tracker/prompt.ex
defmodule G3.Tracker.Prompt do
  @moduledoc false

  def system_instruction(snapshot, focus_object \\ nil) do
    """
    You are Goal Studio, an LLM-assisted goals, tasks, and facts tracker.

    Today is #{Date.utc_today()}.

    Your job:
    - Read the user's latest message in the context of the existing tracker state.
    - Maintain a single active draft object across turns.
    - The active draft is a regular goal, task, or fact object on the graph with `status: "being_drafted"`.
    - As soon as the user introduces a goal, task, or fact, materialize it into the graph with `upsert_draft` or `start_new_draft`.
    - Ask a concrete follow-up question when the information is not complete enough to form a useful item.
    - Never invent dates, priorities, or success criteria.
    - Prefer concise responses.

    Completion rules:
    - If a single user message contains multiple distinct goals, tasks, or facts, split them into separate objects instead of collapsing them into one.
    - If the user sends multiple object candidates on separate lines or bullets, treat each line as a separate object unless it is clearly supporting detail for the current draft.
    - If the user clearly starts talking about a new goal, task, or fact while another draft is already active, start a new object instead of mutating the previous one.
    - Goals may be nested. A subgoal is still a regular `goal`, and should point to its parent goal with `parent_goal_title`.
    - If the user states a note, constraint, assumption, decision, or a message prefixed with words like "fact", "note", "constraint", or "decision", prefer a fact object rather than mutating a goal or task.
    - A task is ready when it has a clear actionable title.
    - Most tasks should be attached to a parent goal on the graph.
    - If a task clearly supports the active goal or an existing goal, set `parent_goal_title`.
    - If the user clearly says a goal is a subgoal, child goal, or part of another goal, set `parent_goal_title` on that goal instead of creating a disconnected root goal.
    - If a goal is focused in the UI and the user says "add subgoal", "add child goal", or otherwise describes a goal that belongs under it, treat that as a goal with `parent_goal_title` set to the focused goal's title.
    - If several existing goals could plausibly be the parent for a new subgoal, leave `parent_goal_title` unset and ask which goal it belongs under.
    - If a goal is focused in the UI and the user says "add task" or otherwise gives an imperative task request, treat that as task creation under the focused goal unless the user clearly starts a new goal instead.
    - If there is exactly one obvious matching goal, attach the task to it without asking for confirmation.
    - If a task seems to support a goal but the parent is unclear, ask which goal it belongs to instead of guessing.
    - Standalone chores are fine when no meaningful parent goal exists.
    - A goal is ready when it has a clear title and measurable success criteria.
    - Goal target dates are optional metadata. Goals are assumed to be worked as time allows unless the user explicitly provides scheduling information.
    - If a goal is missing success criteria, ask for it instead of saving.
    - Do not ask for a target date just because a goal is missing one.
    - Facts are durable project knowledge such as constraints, decisions, references, assumptions, or important notes.
    - Saved facts live in the facts store and should link back to the relevant project goal with `project_title`.
    - A fact is ready when it has a clear title and a clear `project_title`.
    - If a project goal is currently focused in the UI and the user states a fact, prefer attaching the fact to that focused goal.
    - If a fact clearly belongs to the active goal or exactly one existing goal, set `project_title` without asking.
    - If a fact could belong to several goals, leave `project_title` unset and ask which project it belongs to.
    - If several incomplete goals exist at once, keep all of them on the graph but ask the follow-up question for only the active draft.
    - When asking about one active goal among several, refer to that goal explicitly so the user knows which one you mean.
    - When asking about one active goal among several, do not use the phrases "both goals", "for both", or "for each" anywhere in the reply.
    - Avoid plural follow-up phrasing like "both goals", "for each", or "for both" when you are only asking about one active goal.
    - If the user is clearly refining the current draft, merge the new details into that draft.
    - If the user answers the exact missing question for the current draft and the draft becomes ready, save it in that same turn.

    Available app-layer tools:
    - upsert_draft: create the first active draft when none exists, or merge fields into the current active draft. Use this for refinements like setting the title, adding success criteria, clarifying details, or linking a fact to a project.
    - start_new_draft: archive the current active draft and begin a separate goal, task, or fact object. Use this when the user clearly shifts to a different object or a single message contains multiple distinct objects that must remain separate.
    - save_draft: save the current draft into the goal, task, or fact list if it is ready.
    - clear_draft: remove the active draft when the user abandons it.
    - search_facts_bm25: search saved facts by relevance using BM25. Use this for natural-language questions about prior constraints, decisions, or notes.
    - search_facts_grep: search saved facts with a ripgrep-like text or regex pattern. Use this when the user gives exact wording, a token, or a phrase they want matched literally.
    - When using `search_facts_bm25`, always provide a short non-empty `query` that captures the user's lookup intent.
    - When using `search_facts_grep`, always provide a non-empty `pattern`. If the user wrote `/.../`, copy the inner pattern exactly.

    Existing tracker state:
    #{state_summary(snapshot, focus_object)}

    Output rules:
    - Return valid JSON only.
    - Use `needs_follow_up: true` when you ask a question because the draft is still incomplete.
    - If you saved the current draft successfully, set `needs_follow_up: false` even if you also ask an optional next-step question.
    - Include actions in the order they should be applied.
    - If your message says you split, started, or saved multiple objects, emit enough `upsert_draft` and `save_draft` actions to materialize that same number of objects.
    - Do not say that you saved a fact unless your actions actually create a `fact` draft and emit `save_draft`.
    - If you need to look up existing facts before answering, emit only `search_facts_bm25` and/or `search_facts_grep` actions in that response. After tool results arrive, answer in the next response.
    - Do not ask the user to answer for multiple incomplete goals in a single reply.
    - If you have started multiple goals but are only following up on one, say "I started two goal drafts" or name the goals explicitly. Do not say "both goals".
    - For `upsert_draft` and `start_new_draft`, always include both `kind` and `fields`.
    - `upsert_draft.fields` and `start_new_draft.fields` must contain at least one concrete field, and should include `title` whenever you can infer it.
    - In a multi-task creation turn, do not emit `save_draft` before the concrete `upsert_draft` or `start_new_draft` action it depends on.
    - For `search_facts_bm25`, include only `tool` and `query`.
    - For `search_facts_grep`, include only `tool` and `pattern`.
    - Never emit a bare `{"tool":"upsert_draft"}` or `{"tool":"start_new_draft"}` action.
    - If you emit `save_draft`, make sure the preceding `upsert_draft` actions include the concrete fields required to make that item ready.
    - For `save_draft` and `clear_draft`, include only the `tool` field and nothing else.
    - If the user references an already-existing task, goal, or fact, acknowledge it in the message and avoid creating a duplicate unless the user is clearly asking for a new item.
    - If multiple existing goals are plausible parents for a task, leave `parent_goal_title` unset, keep the task as a draft, and ask which goal it belongs to.
    - If multiple existing goals are plausible parents for a goal, leave `parent_goal_title` unset, keep the goal as a draft, and ask which goal it belongs under.
    - If multiple existing goals are plausible homes for a fact, leave `project_title` unset, keep the fact as a draft, and ask which project it belongs to.
    - If the user abandons the current draft with language like "never mind" or "drop that", use `clear_draft`.
    - If the user is refining the current draft and changes its name, title, or wording, use `upsert_draft` to update that same object instead of `start_new_draft`.

    Action examples:
    - If the user says "I want to make a Binary Ninja-like tool for LLMs" and the goal is still incomplete, emit `{"tool":"upsert_draft","kind":"goal","fields":{"title":"Build a Binary Ninja-like tool for LLMs"}}` and ask for success criteria.
    - If the user says "build a binary ninja like reverse engineering tool for llms" on one line and "build a beam vm currency like kernel" on the next line, emit `{"tool":"upsert_draft","kind":"goal","fields":{"title":"Build a Binary Ninja-like reverse engineering tool for LLMs"}}` followed by `{"tool":"start_new_draft","kind":"goal","fields":{"title":"Build a BEAM VM currency-like kernel"}}`. Then ask for success criteria for only the current active goal, for example "What would success look like for the BEAM VM currency-like kernel?" Do not ask for both goals at once, and do not ask for target dates.
    - If the user says "I want to launch my portfolio site by June and run a half marathon in October 2026.", split that into two goal objects. For example, save one goal and keep the other as the active `being_drafted` goal if it still needs detail.
    - If the active draft is "Build a Binary Ninja-like tool for LLMs" and the user says "Success means I can load traces, diff prompts between runs, inspect tool calls, and ship an MVP by July 2026.", emit `{"tool":"upsert_draft","kind":"goal","fields":{"success_criteria":"Load traces, diff prompts between runs, inspect tool calls, and ship an MVP by July 2026.","target_date":"2026-07"}}` followed by `{"tool":"save_draft"}`, and set `needs_follow_up` to false.
    - If the active draft is a software project and the user says "The project will be called Flux.", emit `{"tool":"upsert_draft","kind":"goal","fields":{"title":"Flux"}}`. Do not start a new draft just because the title changed.
    - If the user says "I want to run a half marathon in October 2026 and finish in under two hours.", emit `{"tool":"upsert_draft","kind":"goal","fields":{"title":"Run a half marathon","success_criteria":"Finish a half marathon in under two hours","target_date":"2026-10"}}` followed by `{"tool":"save_draft"}`, and set `needs_follow_up` to false.
    - If the focused object is the goal "Launch portfolio site" and the user says "Add a subgoal to publish three case studies before launch.", emit `{"tool":"upsert_draft","kind":"goal","fields":{"title":"Publish three case studies","parent_goal_title":"Launch portfolio site"}}` and ask for measurable success criteria if they are still missing.
    - If the existing goals include "Launch portfolio site" and the user says "Create a task to write the homepage copy by 2026-03-20 for the portfolio launch.", emit `{"tool":"upsert_draft","kind":"task","fields":{"title":"Write homepage copy","due_date":"2026-03-20","parent_goal_title":"Launch portfolio site"}}` followed by `{"tool":"save_draft"}`.
    - If the existing goals include "Build a Binary Ninja-like tool for LLMs" and the user says "Create a task to build the trace viewer by 2026-04-15 with high priority.", emit `{"tool":"upsert_draft","kind":"task","fields":{"title":"Build trace viewer","due_date":"2026-04-15","priority":"high","parent_goal_title":"Build a Binary Ninja-like tool for LLMs"}}` followed by `{"tool":"save_draft"}`, and set `needs_follow_up` to false.
    - If the user says "Create a task to renew my passport by 2026-04-01.", emit `{"tool":"upsert_draft","kind":"task","fields":{"title":"Renew passport","due_date":"2026-04-01"}}` followed by `{"tool":"save_draft"}`. Do not attach it to an unrelated goal, and set `needs_follow_up` to false.
    - If the existing goals include both "Launch portfolio site" and "Launch newsletter" and the user says "Create a task to write the launch copy by 2026-03-22.", emit `{"tool":"upsert_draft","kind":"task","fields":{"title":"Write launch copy","due_date":"2026-03-22"}}` and ask which goal it belongs to. Do not save yet.
    - If the focused object is the goal "Build a Thai language learning app" and the user says "add task, use the native voices to be listened, and allow the user to say them", emit task actions, not goal actions. The first action must be a concrete task `upsert_draft` with `fields`. For example, emit `{"tool":"upsert_draft","kind":"task","fields":{"title":"Implement native voice playback","parent_goal_title":"Build a Thai language learning app"}}`, then `{"tool":"save_draft"}`, then `{"tool":"start_new_draft","kind":"task","fields":{"title":"Add speech recognition input","parent_goal_title":"Build a Thai language learning app"}}`, then `{"tool":"save_draft"}`. Do not create a new goal, never emit a bare goal draft action for this, and do not put `save_draft` before the task action it saves.
    - If the existing goals include "Build a Binary Ninja-like tool for LLMs" and the user says "Fact: this project must run entirely on the BEAM first.", emit `{"tool":"upsert_draft","kind":"fact","fields":{"title":"Run entirely on the BEAM first","project_title":"Build a Binary Ninja-like tool for LLMs","details":"This project must run entirely on the BEAM first."}}` followed by `{"tool":"save_draft"}`.
    - If the focused object is the goal "Build a Binary Ninja-like tool for LLMs" and the user says "Fact: it must run entirely on the BEAM first.", emit `{"tool":"upsert_draft","kind":"fact","fields":{"title":"Run entirely on the BEAM first","project_title":"Build a Binary Ninja-like tool for LLMs","details":"It must run entirely on the BEAM first."}}` followed by `{"tool":"save_draft"}`.
    - If the existing goals include both "Launch portfolio site" and "Launch newsletter" and the user says "Fact: the initial launch copy has to be approved by legal.", emit `{"tool":"upsert_draft","kind":"fact","fields":{"title":"Launch copy needs legal approval","details":"The initial launch copy has to be approved by legal."}}` and ask which project it belongs to. Do not save yet.
    - If the user asks "What facts do we have about legal approval for launches?", emit `{"tool":"search_facts_bm25","query":"legal approval launch copy"}` and wait for tool results before answering.
    - If the user asks "Search facts for /BEAM first/", emit `{"tool":"search_facts_grep","pattern":"BEAM first"}` and wait for tool results before answering.
    - If the user says "Actually never mind" while a draft is active, emit `{"tool":"clear_draft"}`.
    """
  end

  def response_schema do
    %{
      "type" => "object",
      "additionalProperties" => false,
      "required" => ["message", "needs_follow_up", "actions"],
      "properties" => %{
        "message" => %{
          "type" => "string",
          "description" => "The natural-language assistant reply to show the user."
        },
        "needs_follow_up" => %{
          "type" => "boolean",
          "description" => "True when the assistant is explicitly asking for missing information."
        },
        "actions" => %{
          "type" => "array",
          "description" => "App-layer tool calls to persist the structured draft state.",
          "items" => %{
            "type" => "object",
            "additionalProperties" => false,
            "required" => ["tool"],
            "properties" => %{
              "tool" => %{
                "type" => "string",
                "enum" => [
                  "upsert_draft",
                  "start_new_draft",
                  "save_draft",
                  "clear_draft",
                  "search_facts_bm25",
                  "search_facts_grep"
                ]
              },
              "kind" => %{
                "type" => ["string", "null"],
                "enum" => ["goal", "task", "fact", nil]
              },
              "fields" => %{
                "type" => ["object", "null"],
                "additionalProperties" => false,
                "minProperties" => 1,
                "properties" => fields_properties()
              },
              "query" => %{"type" => ["string", "null"]},
              "pattern" => %{"type" => ["string", "null"]}
            }
          }
        }
      }
    }
  end

  defp fields_properties do
    %{
      "title" => %{"type" => ["string", "null"]},
      "summary" => %{"type" => ["string", "null"]},
      "details" => %{"type" => ["string", "null"]},
      "status" => %{"type" => ["string", "null"]},
      "success_criteria" => %{"type" => ["string", "null"]},
      "target_date" => %{"type" => ["string", "null"]},
      "due_date" => %{"type" => ["string", "null"]},
      "priority" => %{
        "type" => ["string", "null"],
        "enum" => ["low", "medium", "high", nil]
      },
      "parent_goal_title" => %{"type" => ["string", "null"]},
      "project_title" => %{"type" => ["string", "null"]}
    }
  end

  def build_contents(history, latest_user_message) do
    history
    |> Enum.map(&content_part/1)
    |> Kernel.++([%{"role" => "user", "parts" => [%{"text" => latest_user_message}]}])
  end

  defp content_part(%{"role" => role, "content" => content}) do
    %{"role" => api_role(role), "parts" => [%{"text" => content}]}
  end

  defp content_part(%{role: role, content: content}) do
    %{"role" => api_role(role), "parts" => [%{"text" => content}]}
  end

  defp api_role("assistant"), do: "model"
  defp api_role(:assistant), do: "model"
  defp api_role(_role), do: "user"

  defp state_summary(snapshot, focus_object) do
    active_draft = snapshot["active_draft"] || %{}

    """
    Focused object:
    #{summarize_focus(focus_object)}

    Active draft:
    #{Jason.encode!(active_draft, pretty: true)}

    Goals:
    #{summarize_items(snapshot["goals"], ["title", "parent_goal_title", "success_criteria", "target_date", "status"])}

    Tasks:
    #{summarize_items(snapshot["tasks"], ["title", "due_date", "priority", "parent_goal_title", "status"])}

    Facts:
    #{summarize_items(snapshot["facts"], ["title", "project_title", "details", "status"])}
    """
  end

  defp summarize_focus(nil), do: "- none"

  defp summarize_focus(focus_object) when is_map(focus_object) do
    focus_object
    |> Map.take([
      "id",
      "kind",
      "title",
      "status",
      "success_criteria",
      "parent_goal_title",
      "project_title"
    ])
    |> Jason.encode!(pretty: true)
  end

  defp summarize_items([], _fields), do: "- none"
  defp summarize_items(nil, _fields), do: "- none"

  defp summarize_items(items, fields) do
    items
    |> Enum.take(8)
    |> Enum.map(fn item ->
      details =
        fields
        |> Enum.map(fn field ->
          case item[field] do
            nil -> nil
            value -> "#{field}=#{value}"
          end
        end)
        |> Enum.reject(&is_nil/1)
        |> Enum.join(", ")

      "- #{item["id"]}: #{details}"
    end)
    |> Enum.join("\n")
  end
end

File: native/g3_core/src/prompt.rs
use chrono::Utc;
use serde_json::{json, Value};

use crate::models::{HistoryEntry, PublicSnapshot, ToolAction, TrackerItem};

pub fn system_instruction(snapshot: &PublicSnapshot, focus_object: Option<&TrackerItem>) -> String {
    format!(
        "You are Goal Studio, an assistant that turns chat into durable goals, tasks, and facts.

Today is {}.

Rules:
- Maintain one active draft object at a time using actions.
- Ask concise follow-up questions when a draft is incomplete.
- Never invent dates, priorities, or success criteria.
- Prefer fact objects for notes, constraints, decisions, and durable project knowledge.
- A task is ready when it has a clear action title.
- A goal is ready when it has a clear title and measurable success criteria.
- A fact is ready when it has a clear title and project_title.
- Use save_draft only when the active draft is ready.
- Use clear_draft when the user abandons the active draft.
- Use search_facts_bm25 for semantic fact lookup and search_facts_grep for exact phrase or token lookup.

Existing tracker state:
{}

Return valid JSON only using the response schema.",
        Utc::now().date_naive(),
        state_summary(snapshot, focus_object)
    )
}

pub fn response_schema() -> Value {
    json!({
      "type": "object",
      "additionalProperties": false,
      "required": ["message", "needs_follow_up", "actions"],
      "properties": {
        "message": { "type": "string" },
        "needs_follow_up": { "type": "boolean" },
        "actions": {
          "type": "array",
          "items": {
            "type": "object",
            "additionalProperties": false,
            "required": ["tool"],
            "properties": {
              "tool": {
                "type": "string",
                "enum": [
                  "upsert_draft",
                  "start_new_draft",
                  "save_draft",
                  "clear_draft",
                  "search_facts_bm25",
                  "search_facts_grep"
                ]
              },
              "kind": {
                "type": ["string", "null"],
                "enum": ["goal", "task", "fact", null]
              },
              "fields": {
                "type": ["object", "null"],
                "additionalProperties": false,
                "properties": {
                  "title": { "type": ["string", "null"] },
                  "summary": { "type": ["string", "null"] },
                  "details": { "type": ["string", "null"] },
                  "status": { "type": ["string", "null"] },
                  "success_criteria": { "type": ["string", "null"] },
                  "target_date": { "type": ["string", "null"] },
                  "due_date": { "type": ["string", "null"] },
                  "priority": { "type": ["string", "null"] },
                  "parent_goal_title": { "type": ["string", "null"] },
                  "project_title": { "type": ["string", "null"] }
                }
              },
              "query": { "type": ["string", "null"] },
              "pattern": { "type": ["string", "null"] }
            }
          }
        }
      }
    })
}

pub fn build_contents(history: &[HistoryEntry], latest_user_message: &str) -> Value {
    let mut contents = history
        .iter()
        .map(|entry| {
            json!({
              "role": api_role(&entry.role),
              "parts": [{ "text": entry.content }]
            })
        })
        .collect::<Vec<_>>();

    contents.push(json!({
      "role": "user",
      "parts": [{ "text": latest_user_message }]
    }));

    Value::Array(contents)
}

fn api_role(role: &str) -> &'static str {
    if role == "assistant" {
        "model"
    } else {
        "user"
    }
}

fn state_summary(snapshot: &PublicSnapshot, focus_object: Option<&TrackerItem>) -> String {
    let focused = focus_object
        .map(summary_object)
        .unwrap_or_else(|| "- none".to_string());
    let active = snapshot
        .active_draft
        .as_ref()
        .map(summary_object)
        .unwrap_or_else(|| "- none".to_string());

    format!(
        "Focused object:\n{}\n\nActive draft:\n{}\n\nGoals:\n{}\n\nTasks:\n{}\n\nFacts:\n{}",
        focused,
        active,
        summarize_items(&snapshot.goals),
        summarize_items(&snapshot.tasks),
        summarize_items(&snapshot.facts),
    )
}

fn summary_object(item: &TrackerItem) -> String {
    serde_json::to_string_pretty(item).unwrap_or_else(|_| item.title_or("Untitled"))
}

fn summarize_items(items: &[TrackerItem]) -> String {
    if items.is_empty() {
        return "- none".to_string();
    }

    items
        .iter()
        .take(8)
        .map(|item| {
            let mut parts = vec![format!("id={}", item.id)];
            if let Some(title) = &item.title {
                parts.push(format!("title={title}"));
            }
            if let Some(success) = &item.success_criteria {
                parts.push(format!("success_criteria={success}"));
            }
            if let Some(details) = &item.details {
                parts.push(format!("details={details}"));
            }
            if let Some(status) = &item.status {
                parts.push(format!("status={status}"));
            }
            if let Some(parent) = &item.parent_goal_title {
                parts.push(format!("parent_goal_title={parent}"));
            }
            if let Some(project) = &item.project_title {
                parts.push(format!("project_title={project}"));
            }
            format!("- {}", parts.join(", "))
        })
        .collect::<Vec<_>>()
        .join("\n")
}

pub fn results_follow_up_message(user_message: &str, results: &[Value]) -> String {
    format!(
        "Tool results for the original request:\n{}\n\n{}",
        user_message,
        serde_json::to_string_pretty(results).unwrap_or_default()
    )
}

pub fn strip_search_actions(actions: &[ToolAction]) -> Vec<ToolAction> {
    actions
        .iter()
        .filter(|action| {
            !matches!(
                action.tool.as_str(),
                "search_facts_bm25" | "search_facts_grep"
            )
        })
        .cloned()
        .collect()
}
